File: sdk_v2/js/src/foundryLocalManager.ts
// Singleton enforcement: the underlying `foundry_local::Manager` is a process singleton — the C++ wrapper throws
// if a second one is created. We do NOT duplicate that check in JS; the user-facing failure mode is a tagged
// `FoundryLocalError` from the native addon. `create()` and `createAsync()` are factory wrappers around the
// constructor — the native layer is the source of truth for instance identity.

import { type Catalog, wrapNativeCatalog } from "./catalog.js";
import { FOUNDRY_LOCAL_CONFIG_KEYS, type FoundryLocalConfig } from "./configuration.js";
import {
  type NativeManager,
  configureNativeLoader,
  getAddon,
  getPreloadedLibraryPath,
} from "./detail/native.js";
import type { EpDownloadResult, EpInfo } from "./types.js";

// A native Manager holds process-global native resources. Dispose the live
// Manager on process exit so native teardown happens at a deterministic point
// rather than being left entirely to environment finalizers. `beforeExit`
// covers a natural event-loop drain; `exit` covers callers who invoke
// `process.exit()` themselves. Both are idempotent. The native layer permits
// only one live Manager at a time, so a single reference is enough.
let liveManager: FoundryLocalManager | undefined;
let exitHandlersInstalled = false;

function disposeLiveManager(): void {
  const manager = liveManager;
  if (manager === undefined) {
    return;
  }
  try {
    manager.dispose();
  } catch {
    // Best-effort: a dispose failure must not block process exit.
  }
}

function installExitHandlersOnce(): void {
  if (exitHandlersInstalled) {
    return;
  }
  exitHandlersInstalled = true;
  process.on("beforeExit", () => {
    disposeLiveManager();
  });
  process.on("exit", () => {
    disposeLiveManager();
  });
}

export class FoundryLocalManager {
  readonly #native: NativeManager;
  #catalog: Catalog | undefined;
  #urls: string[] = [];

  constructor(config: FoundryLocalConfig) {
    if (config === null || config === undefined) {
      throw new TypeError("Configuration must be provided.");
    }
    if (typeof config.appName !== "string" || config.appName.trim() === "") {
      throw new TypeError("appName must be set to a valid application name.");
    }
    // Reject typos like `cachePath` (vs. `modelCacheDir`). The native layer would otherwise silently ignore
    // them, sending the user down a confusing debug path (the wrong cache dir, wrong service URL, etc.).
    const unknownKeys = Object.keys(config).filter(
      (key) => !FOUNDRY_LOCAL_CONFIG_KEYS.has(key as keyof FoundryLocalConfig),
    );
    if (unknownKeys.length > 0) {
      const allowed = [...FOUNDRY_LOCAL_CONFIG_KEYS].sort().join(", ");
      throw new TypeError(
        `Unknown FoundryLocalConfig ${unknownKeys.length === 1 ? "property" : "properties"}: ${unknownKeys.join(", ")}. Allowed: ${allowed}.`,
      );
    }
    // Apply libraryPath (if any) before triggering addon load. If the addon
    // is already loaded, only re-applying the same path is a silent no-op;
    // any other path is rejected with a clear message.
    if (config.libraryPath !== undefined && config.libraryPath !== "") {
      const already = getPreloadedLibraryPath();
      if (already !== undefined && already !== config.libraryPath) {
        throw new Error(
          "libraryPath cannot be changed after the native addon has been loaded. Subsequent FoundryLocalManager instances must use the same libraryPath (or omit it).",
        );
      }
      if (already === undefined) {
        try {
          configureNativeLoader({ libraryPath: config.libraryPath });
        } catch (err) {
          if (err instanceof Error) {
            throw new Error(`Failed to configure native loader: ${err.message}`, { cause: err });
          }
          throw err;
        }
      }
    }
    this.#native = new (getAddon().Manager)(config);
    liveManager = this;
    installExitHandlersOnce();
  }

  /**
   * Construct a `FoundryLocalManager`. The native layer is the source of truth for the process singleton —
   * calling `create()` more than once throws a tagged `FoundryLocalError` from the C++ wrapper.
   *
   * The native constructor does no I/O, so there is no difference in blocking behaviour between `create` and
   * `createAsync` other than returning a Promise.
   */
  static create(config: FoundryLocalConfig): FoundryLocalManager {
    return new FoundryLocalManager(config);
  }

  /** Async factory variant. */
  static createAsync(config: FoundryLocalConfig): Promise<FoundryLocalManager> {
    return Promise.resolve(new FoundryLocalManager(config));
  }

  /** The model catalog. Lazily wraps the native handle and is cached. */
  get catalog(): Catalog {
    if (this.#catalog === undefined) {
      this.#catalog = wrapNativeCatalog(this.#native.getCatalog());
    }
    return this.#catalog;
  }

  /** URLs the embedded web service is bound to. Empty when not running. */
  get urls(): string[] {
    return this.#urls;
  }

  /** True when the embedded web service is currently running. */
  get isWebServiceRunning(): boolean {
    return this.#urls.length > 0;
  }

  /** Start the embedded web service. Updates `urls`. */
  startWebService(): void {
    this.#native.startWebService();
    this.#urls = this.#native.getWebServiceEndpoints();
  }

  /** Stop the embedded web service. Clears `urls`. */
  stopWebService(): void {
    if (this.#urls.length === 0) {
      return;
    }
    this.#native.stopWebService();
    this.#urls = [];
  }

  /** Discover available execution providers and their registration status. */
  discoverEps(): EpInfo[] {
    return this.#native.discoverEps();
  }

  /** Download and register execution providers. */
  downloadAndRegisterEps(): Promise<EpDownloadResult>;
  /** Download and register specific execution providers by name. */
  downloadAndRegisterEps(names: string[]): Promise<EpDownloadResult>;
  /** Download and register all EPs, with a progress callback. */
  downloadAndRegisterEps(progressCallback: (epName: string, percent: number) => void): Promise<EpDownloadResult>;
  /** Download and register named EPs, with a progress callback. */
  downloadAndRegisterEps(
    names: string[],
    progressCallback: (epName: string, percent: number) => void,
  ): Promise<EpDownloadResult>;
  async downloadAndRegisterEps(
    namesOrCallback?: string[] | ((epName: string, percent: number) => void),
    progressCallback?: (epName: string, percent: number) => void,
  ): Promise<EpDownloadResult> {
    let names: string[] | undefined;
    let cb: ((epName: string, percent: number) => void) | undefined;
    if (typeof namesOrCallback === "function") {
      cb = namesOrCallback;
    } else if (Array.isArray(namesOrCallback)) {
      names = namesOrCallback;
      cb = progressCallback;
    } else if (progressCallback !== undefined) {
      cb = progressCallback;
    }

    // The native API throws on failure rather than returning a pass/fail result envelope. We capture which EPs
    // the native call asked to register and report them back as `registeredEps` on success; on failure we
    // re-throw the tagged FoundryLocalError unchanged.
    const requested = names ?? [];
    await this.#native.downloadAndRegisterEps(names, cb);

    return {
      success: true,
      status: requested.length > 0 ? "Requested EPs registered" : "All providers registered",
      registeredEps: requested,
      failedEps: [],
    };
  }

  /** Whether an EP download/registration operation is currently in progress. */
  isEpDownloadInProgress(): boolean {
    return this.#native.isEpDownloadInProgress();
  }

  /** Begin graceful shutdown. Idempotent. */
  shutdown(): void {
    this.#native.shutdown();
  }

  /** Whether `shutdown()` has been called. */
  get isShutdownRequested(): boolean {
    return this.#native.isShutdownRequested();
  }

  /** True if `dispose()` has been called on this manager. */
  get disposed(): boolean {
    return this.#native.isDisposed();
  }

  /**
   * Release the underlying native `foundry_local::Manager`. Idempotent. After disposal, every other instance method
   * (and any method on a `Catalog` or `Model` obtained through this manager) throws a `FoundryLocalError`.
   */
  dispose(): void {
    if (liveManager === this) {
      liveManager = undefined;
    }
    this.#native.dispose();
    this.#catalog = undefined;
    this.#urls = [];
  }

  [Symbol.dispose](): void {
    this.dispose();
  }
}

File: sdk_v2/rust/src/foundry_local_manager.rs
//! Top-level entry point for the Foundry Local SDK.
//!
//! [`FoundryLocalManager`] initialises the native core library, provides access
//! to the model [`Catalog`], and can start / stop the local web service. While a
//! handle is alive it is shared process-wide (see [`FoundryLocalManager::create`]).

use std::sync::atomic::AtomicBool;
use std::sync::{Arc, Mutex, OnceLock, Weak};

use tokio::sync::Mutex as AsyncMutex;

use crate::catalog::Catalog;
use crate::configuration::{FoundryLocalConfig, Logger};
use crate::detail::api::Api;
use crate::detail::manager::{EpProgressCallback, NativeManager};
use crate::detail::task::spawn_blocking;
use crate::error::{FoundryLocalError, Result};
use crate::types::{EpDownloadResult, EpInfo};

/// Process-wide weak handle to the live manager.
///
/// Holds a [`Weak`] so the global never keeps the manager alive past its last
/// strong reference: when the final [`Arc`] returned by
/// [`FoundryLocalManager::create`] is dropped, the native manager is torn down
/// deterministically via [`Drop`] — while the ORT runtime is still alive and
/// before the library's C++ static destructors run.
///
/// Wrapped in a [`OnceLock`] (rather than a `const` `Mutex::new`) to keep the
/// crate compatible with its minimum supported Rust version.
static INSTANCE: OnceLock<Mutex<Weak<FoundryLocalManager>>> = OnceLock::new();

/// The lazily-initialised slot holding the shared-instance weak handle.
fn instance_slot() -> &'static Mutex<Weak<FoundryLocalManager>> {
    INSTANCE.get_or_init(|| Mutex::new(Weak::new()))
}

/// Primary entry point for interacting with Foundry Local.
///
/// Obtain a handle with [`FoundryLocalManager::create`]. While at least one
/// handle is alive, every caller shares the same instance.
pub struct FoundryLocalManager {
    native: Arc<NativeManager>,
    catalog: Catalog,
    urls: Mutex<Vec<String>>,
    /// Serialises web-service start/stop so concurrent callers cannot race on
    /// the native lifecycle state (which the C++ core mutates without a lock).
    web_service_lock: AsyncMutex<()>,
    /// Application logger (stub — not yet wired into the native core).
    _logger: Option<Box<dyn Logger>>,
}

type EpDownloadProgressCallback = Box<dyn FnMut(&str, f64) + Send + 'static>;

/// Builder for configuring and running execution provider downloads.
pub struct EpDownloadBuilder<'a> {
    manager: &'a FoundryLocalManager,
    names: Option<Vec<String>>,
    progress_callback: Option<EpDownloadProgressCallback>,
    cancel_flag: Option<Arc<AtomicBool>>,
}

impl<'a> EpDownloadBuilder<'a> {
    fn new(manager: &'a FoundryLocalManager) -> Self {
        Self {
            manager,
            names: None,
            progress_callback: None,
            cancel_flag: None,
        }
    }

    /// Download only the named execution providers.
    pub fn names<I, S>(mut self, names: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        self.names = Some(names.into_iter().map(Into::into).collect());
        self
    }

    /// Report per-EP download progress as `(ep_name, percent)`.
    pub fn progress<F>(mut self, callback: F) -> Self
    where
        F: FnMut(&str, f64) + Send + 'static,
    {
        self.progress_callback = Some(Box::new(callback));
        self
    }

    /// Cancel the download when `cancel_flag` is set to `true`.
    pub fn cancel(mut self, cancel_flag: Arc<AtomicBool>) -> Self {
        self.cancel_flag = Some(cancel_flag);
        self
    }

    /// Run the configured execution provider download.
    pub async fn run(self) -> Result<EpDownloadResult> {
        self.manager
            .download_and_register_eps_impl(self.names, self.progress_callback, self.cancel_flag)
            .await
    }
}

impl FoundryLocalManager {
    /// Initialise the SDK and return a shared handle to the manager.
    ///
    /// While at least one returned [`Arc`] is alive, every call returns the
    /// **same** instance (a process-wide singleton) and the `config` passed to
    /// later calls is ignored. The native manager is torn down once every handle
    /// derived from it is gone — this `Arc`, plus any [`Model`](crate::Model),
    /// client, or session it produced, each of which keeps the native manager
    /// alive so handles can safely outlive this `Arc`. A subsequent call then
    /// builds a fresh instance from the new `config`.
    ///
    /// Teardown runs via [`Drop`] when the final handle is released — not via a
    /// process-exit hook — so the native manager (and its EP unregistration)
    /// shuts down while the engine / ORT runtime is still alive. This matches
    /// the C++ SDK's local-`Manager` semantics and avoids the WebGPU
    /// `ReleaseEpFactory` teardown throw (ORT #29206).
    pub fn create(config: FoundryLocalConfig) -> Result<Arc<Self>> {
        // Hold the lock across initialisation so only one thread builds the
        // instance; concurrent callers then observe and share it. A poisoned
        // lock is recoverable: the guarded `Weak` is valid regardless of panics.
        let mut slot = instance_slot()
            .lock()
            .unwrap_or_else(|poisoned| poisoned.into_inner());

        // Reuse the live instance if one already exists.
        if let Some(existing) = slot.upgrade() {
            return Ok(existing);
        }

        let mut config = config;
        let api = Arc::new(Api::load(config.library_path_ref())?);
        let logger = config.take_logger();
        let native_config = config.build_native(&api)?;

        let native = Arc::new(NativeManager::create(
            Arc::clone(&api),
            native_config.as_ptr(),
        )?);

        let catalog_ptr = native.catalog_ptr()?;
        let catalog = Catalog::new(Arc::clone(&api), catalog_ptr, Arc::clone(&native))?;

        let manager = Arc::new(FoundryLocalManager {
            native,
            catalog,
            urls: Mutex::new(Vec::new()),
            web_service_lock: AsyncMutex::new(()),
            _logger: logger,
        });

        // Record a weak reference so future calls share this instance without
        // keeping it alive past the caller's last strong handle.
        *slot = Arc::downgrade(&manager);
        Ok(manager)
    }

    /// Access the model catalog.
    pub fn catalog(&self) -> &Catalog {
        &self.catalog
    }

    /// Begin a graceful shutdown of the local engine.
    ///
    /// Stops the web service, prevents new model loads, stops existing
    /// sessions, and unloads models. Idempotent and safe to call from any
    /// thread.
    ///
    /// Calling this is optional: the native manager is released automatically
    /// when the last handle is dropped. Use it when you want to deterministically
    /// wind the engine down before releasing the handle. After calling
    /// `shutdown`, the manager should not be used for further inference.
    pub fn shutdown(&self) -> Result<()> {
        self.native.shutdown()
    }

    /// URLs that the local web service is listening on.
    ///
    /// Empty until [`Self::start_web_service`] has been called.
    pub fn urls(&self) -> Result<Vec<String>> {
        let lock = self.urls.lock().map_err(|_| FoundryLocalError::Internal {
            reason: "Failed to acquire urls lock".into(),
        })?;
        Ok(lock.clone())
    }

    /// Start the local web service.
    ///
    /// The listening URLs are stored internally and can be retrieved via
    /// [`Self::urls`] after this method returns.
    pub async fn start_web_service(&self) -> Result<()> {
        let _guard = self.web_service_lock.lock().await;
        let native = Arc::clone(&self.native);
        let urls = spawn_blocking(move || {
            native.web_service_start()?;
            native.web_service_urls()
        })
        .await?;
        *self.urls.lock().map_err(|_| FoundryLocalError::Internal {
            reason: "Failed to acquire urls lock".into(),
        })? = urls;
        Ok(())
    }

    /// Stop the local web service.
    pub async fn stop_web_service(&self) -> Result<()> {
        let _guard = self.web_service_lock.lock().await;
        let native = Arc::clone(&self.native);
        spawn_blocking(move || native.web_service_stop()).await?;
        self.urls
            .lock()
            .map_err(|_| FoundryLocalError::Internal {
                reason: "Failed to acquire urls lock".into(),
            })?
            .clear();
        Ok(())
    }

    /// Discover available execution providers and their registration status.
    pub fn discover_eps(&self) -> Result<Vec<EpInfo>> {
        self.native.discover_eps()
    }

    /// Download and register execution providers.
    ///
    /// If `names` is `None` or empty, all available EPs are downloaded.
    /// Otherwise only the named EPs are downloaded and registered.
    pub async fn download_and_register_eps(
        &self,
        names: Option<&[&str]>,
    ) -> Result<EpDownloadResult> {
        let names = names.map(|n| n.iter().map(|s| s.to_string()).collect::<Vec<_>>());
        self.download_and_register_eps_impl(names, None, None).await
    }

    /// Download and register execution providers, reporting per-EP progress.
    ///
    /// If `names` is `None` or empty, all available EPs are downloaded.
    /// Otherwise only the named EPs are downloaded and registered.
    ///
    /// `progress_callback` receives `(ep_name, percent)` where `percent`
    /// ranges from 0.0 to 100.0 as each EP downloads.
    pub async fn download_and_register_eps_with_progress<F>(
        &self,
        names: Option<&[&str]>,
        progress_callback: F,
    ) -> Result<EpDownloadResult>
    where
        F: FnMut(&str, f64) + Send + 'static,
    {
        let names = names.map(|n| n.iter().map(|s| s.to_string()).collect::<Vec<_>>());
        self.download_and_register_eps_impl(names, Some(Box::new(progress_callback)), None)
            .await
    }

    /// Configure and run execution provider downloads with a builder.
    ///
    /// Use this for call sites that need names, progress, cancellation, or
    /// future download options.
    pub fn download_and_register_eps_builder(&self) -> EpDownloadBuilder<'_> {
        EpDownloadBuilder::new(self)
    }

    async fn download_and_register_eps_impl(
        &self,
        names: Option<Vec<String>>,
        progress_callback: Option<EpDownloadProgressCallback>,
        cancel_flag: Option<Arc<AtomicBool>>,
    ) -> Result<EpDownloadResult> {
        let native = Arc::clone(&self.native);

        // Snapshot requested EP names (default: all discoverable). A discovery
        // failure must propagate rather than collapse into an empty set, which
        // would silently invoke the "download all" path and report success for
        // zero requested providers.
        let requested: Vec<String> = match &names {
            Some(n) if !n.is_empty() => n.clone(),
            _ => native.discover_eps()?.into_iter().map(|e| e.name).collect(),
        };

        let (message, after) = spawn_blocking(move || {
            let name_refs: Option<Vec<&str>> = names
                .as_ref()
                .map(|n| n.iter().map(String::as_str).collect());
            let progress: Option<EpProgressCallback> =
                progress_callback.map(|cb| cb as EpProgressCallback);
            let message =
                native.download_and_register_eps(name_refs.as_deref(), progress, cancel_flag)?;
            // Re-query registration state to synthesise the per-EP result; an
            // observation failure here must surface, not be masked as empty.
            let after = native.discover_eps()?;
            Ok::<(Option<String>, Vec<EpInfo>), FoundryLocalError>((message, after))
        })
        .await?;

        let registered_eps: Vec<String> = requested
            .iter()
            .filter(|name| after.iter().any(|e| &e.name == *name && e.is_registered))
            .cloned()
            .collect();
        let failed_eps: Vec<String> = requested
            .iter()
            .filter(|name| !registered_eps.contains(*name))
            .cloned()
            .collect();

        let success = message.is_none() && failed_eps.is_empty();
        let status = match &message {
            None => "All requested execution providers were registered successfully.".to_string(),
            Some(msg) if msg.is_empty() => {
                "One or more execution providers failed to register.".to_string()
            }
            Some(msg) => msg.clone(),
        };

        let result = EpDownloadResult {
            success,
            status,
            registered_eps,
            failed_eps,
        };

        // Invalidate the catalog cache if any EP was newly registered so the next
        // access re-fetches models with the updated set of available EPs.
        if result.success || !result.registered_eps.is_empty() {
            let _ = self.catalog.update_models().await;
        }

        Ok(result)
    }
}
